src/mame/bfm/bfm_cobra3.cpp

```diff
@@ -0,0 +1,592 @@
+	required_device<palette_device> m_palette;
+	required_device<ramdac_device> m_ramdac;
+	required_device<scc66470_device> m_scc66470;
+	// required_ioport_array<4> m_strobein;
+	optional_device<meters_device> m_meters;
+	output_finder<256> m_lamps;
+	required_device<nscsi_bus_device> m_scsibus;
+	required_device<ncr5380_device> m_scsic;
+
+
+	virtual void machine_start() override ATTR_COLD;
+	static void scsi_devices(device_slot_interface &device);
+	void dma1_drq(int state);
+	void scc66470_map(address_map &map);
+	void scc66470_irq(int state);
+
+};
+
+uint16_t bfm_cobra3_state::bfm_cobra3_mem_r(offs_t offset, uint16_t mem_mask)
+{
+	int pc = m_maincpu->pc();
+	int cs = m_maincpu->get_cs(offset * 2);
+
+	switch ( cs )
+	{
+		case 1: //ROM
+			return m_cpuregion[offset & 0x7ffff];
+		case 2://RAM
+			return m_mainram[offset & 0x1fff];
+
+		case 3: //I/O 00ac0000-00ac07ff
+			{
+				offset &= 0x7ff;
+				offs_t cs_addr_8_11 = (offset * 2) & 0xf00;
+
+				switch(cs_addr_8_11)
+				{
+					case 0x300: //ac0300
+						if(ACCESSING_BITS_0_7)
+						{
+							return m_ymz->read(offset & 1);
+						}
+						break;
+
+					case 0x400:
+						//input reads
+						break;
+
+					case 0x500: //SCSI DMA
+						if(ACCESSING_BITS_8_15)
+						{
+							return m_scsic->dma_r() << 8;
+						}
+						break;
+
+					case 0x600: //Looks like the RAMDAC hookup
+						if(ACCESSING_BITS_0_7)
+						{
+							if((offset & 7) == 1)
+							{
+								return m_ramdac->pal_r();
+							}
+						}
+						break;
+						
+					default:
+						logerror("%08x maincpu read access offset %08x mem_mask %08x cs %d\n", pc, offset*4, mem_mask, cs);
+						break;
+				}
+			}
+			break;
+
+		case 4: //0x00b10000 SCSI controller
+				if(ACCESSING_BITS_8_15)
+				{
+					return(m_scsic->read(offset & 0x0f)<<8);
+				}
+				break;
+
+		default:
+				logerror("%08x maincpu read access offset %08x mem_mask %08x cs %d\n", pc, offset*4, mem_mask, cs);
+	}
+
+	return 0x0000;
+}
+
+void bfm_cobra3_state::bfm_cobra3_mem_w(offs_t offset, uint16_t data, uint16_t mem_mask)
+{
+	int pc = m_maincpu->pc();
+	int cs = m_maincpu->get_cs(offset * 2);
+
+	switch (cs)
+	{
+		case 1://ROM, shouldn't write here?
+				logerror("%08x maincpu write access(1) offset %08x data %08x mem_mask %08x cs %d\n", pc, offset*4, data, mem_mask, cs);
+				break;
+			
+		case 2://RAM
+				COMBINE_DATA(&m_mainram[offset & 0x1fff]);
+				break;
+			
+		case 3: //00ac0000-00ac07ff I/O
+			{
+				offset &= 0x7ff;
+				offs_t cs_addr_8_11 = (offset * 2) & 0xf00;
+
+				switch(cs_addr_8_11)
+				{
+					case 0x000:
+						// lamps0_15_w(data);
+						break;
+
+					case 0x100:
+						if(ACCESSING_BITS_8_15)
+						{
+							//coin lockout and optional vfd
+						}
+						break;
+
+					case 0x200:
+						//volume, watchdog and other stuff ?
+						break;
+
+					case 0x300:
+						if(ACCESSING_BITS_0_7)
+						{
+							m_ymz->write(offset & 1, data);
+						}
+						break;
+
+					case 0x500://SCSI DMA
```

Comment:
Spacing, again. Please try to maintain a space between the comment marker and the comment text.